File: rust/src/219_contains_nearby_duplicate/contains_nearby_duplicate.rs
/*
 * @Date: 2022-01-19 01:47:57
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-01-19 01:51:57
 */

pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
    use std::collections::HashSet;
    let mut cache = HashSet::new();

    for i in 0..nums.len() {
        if i as i32 - k > 0 {
            cache.remove(&nums[i - k as usize - 1]);
        }
        if !cache.insert(nums[i]) {
            return true;
        }
    }
    false
}

fn main() {
    assert_eq!(contains_nearby_duplicate(vec![1, 2, 3, 1], 3), true);
    assert_eq!(contains_nearby_duplicate(vec![1, 0, 1, 1], 1), true);
    assert_eq!(contains_nearby_duplicate(vec![1, 2, 3, 1, 2, 3], 2), false);
}

File: cpp/219_contains_nearby_duplicate/contains_nearby_duplicate.cpp
/*
 * @Date: 2022-01-19 01:47:54
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2022-01-19 01:52:56
 */

#include <cassert>
#include <tuple>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
 public:
  bool containsNearbyDuplicate(vector<int> nums, int k) {
    unordered_set<int> s;
    int length = nums.size();
    for (int i = 0; i < length; i++) {
      if (i > k) {
        s.erase(nums[i - k - 1]);
      }
      if (s.count(nums[i])) {
        return true;
      }
      s.emplace(nums[i]);
    }
    return false;
  }
};

int main() {
  vector<tuple<vector<int>, int, bool>> tests{
      {{1, 2, 3, 1}, 3, true},
      {{1, 0, 1, 1}, 1, true},
      {{1, 2, 3, 1, 2, 3}, 2, false},
  };
  

  for (auto &[nums, k, ans] : tests) {
    assert(Solution().containsNearbyDuplicate(nums, k) == ans);
  }
}